front/lib/resources/file_resource.ts

```diff
@@ -0,0 +1,228 @@
+    }
+  }
+
+  // Status logic.
+
+  private async update(
+    blob: Partial<CreationAttributes<FileModel>>
+  ): Promise<void> {
+    const [, affectedRows] = await this.model.update(blob, {
+      where: {
+        id: this.id,
+      },
+      returning: true,
+    });
+
+    // Update the resource to reflect the new status.
+    Object.assign(this, affectedRows[0].get());
+  }
+
+  async markAsFailed(): Promise<void> {
+    return this.update({ status: "failed" });
+  }
+
+  async markAsReady(): Promise<void> {
+    return this.update({ status: "ready" });
+  }
+
+  get isReady(): boolean {
+    return this.status === "ready";
+  }
+
+  get isCreated(): boolean {
+    return this.status === "created";
+  }
+
+  get isFailed(): boolean {
+    return this.status === "failed";
+  }
+
+  // Cloud storage logic.
+
+  getPublicUrl(auth: Authenticator): string {
+    // TODO(2024-07-01 flav) Remove once we introduce AuthenticatorWithWorkspace.
+    const owner = auth.workspace();
+    if (!owner) {
+      throw new Error("Unexpected unauthenticated call to `getPublicUrl`");
+    }
+
+    return `${config.getAppUrl()}/api/w/${owner.sId}/files/${this.sId}`;
+  }
+
+  getCloudStoragePath(auth: Authenticator, version: FileVersion): string {
+    // TODO(2024-07-01 flav) Remove once we introduce AuthenticatorWithWorkspace.
+    const owner = auth.workspace();
+    if (!owner) {
+      throw new Error("Unexpected unauthenticated call to `getUploadUrl`");
+    }
+
+    return `files/w/${owner.sId}/${this.sId}/${version}`;
+  }
+
+  async getSignedUrlForDownload(
+    auth: Authenticator,
+    version: FileVersion
+  ): Promise<string> {
+    return getPrivateUploadBucket().getSignedUrl(
+      this.getCloudStoragePath(auth, version),
+      {
+        // Since we redirect, the use is immediate so expiry can be short.
+        expirationDelay: 10 * 1000,
+        promptSaveAs: this.fileName ?? `dust_${this.sId}`,
+      }
+    );
+  }
+
+  // Stream logic.
+
+  getWriteStream(auth: Authenticator, version: FileVersion): Writable {
+    return getPrivateUploadBucket()
+      .file(this.getCloudStoragePath(auth, version))
+      .createWriteStream({
+        resumable: false,
+        gzip: true,
+      });
+  }
+
+  getReadStream(auth: Authenticator, version: FileVersion): Readable {
+    return getPrivateUploadBucket()
+      .file(this.getCloudStoragePath(auth, version))
+      .createReadStream();
+  }
+
+  // Serialization logic.
+
+  toJSON(auth: Authenticator): FileType {
+    const blob: FileType = {
+      contentType: this.contentType,
+      fileName: this.fileName,
+      fileSize: this.fileSize,
+      id: this.sId,
+      status: this.status,
+      useCase: this.useCase,
+    };
+
+    if (this.isCreated) {
+      blob.uploadUrl = this.getPublicUrl(auth);
+    } else if (this.isReady) {
+      blob.downloadUrl = this.getPublicUrl(auth);
+    }
+
+    return blob;
+  }
+}
+
+function makeDustFileId(): FileId {
```

Comment:
Let's modify generateModelSId to take an optional prefix instead?

Reply:
Wanted to do it at first, but was not sure we were okay with it 🔥 !